lib/src/async_expand.dart

```diff
@@ -29,12 +30,18 @@ extension AsyncExpand<T> on Stream<T> {
   /// If the source stream is a broadcast stream, the result will be as well,
   /// regardless of the types of streams created by [convert]. In this case,
   /// some care should be taken:
-  /// -  If [convert] returns a single subscription stream it may be listened to
-  /// and never canceled.
-  /// -  For any period of time where there are no listeners on the result
-  /// stream, any sub streams from previously emitted events will be ignored,
-  /// regardless of whether they emit further events after a listener is added
-  /// back.
+  /// -  The source stream will be immediately listened to, regardless of
```

Comment:
I made this change as a further step towards the principles discussed in https://github.com/dart-lang/stream_transform/pull/148#discussion_r895434468

The problem with not listening is that it becomes observable _when_ the first listener gets added on the stream. If we want the observable behavior that listener see on the stream to be independent of each other I think we have to listen on the outer stream immediately.

Reply:
Ah, I see the other comment about using a multi stream to preserve isolation between listeners... I like that idea.

I'll rework this.

Reply:
0That too was a problem, caused by the returned stream being a broadcast stream.

We don't have to listen to the outer stream immediately, if the individual listeners are completely separate computations. If the resulting stream is *never* a broadcast stream, then we don't have to worry about each listener seeing the same events at the same time, and it won't matter who listens first.

We can make the new stream a single-subscription stream. If you want to listen twice, wrap the original stream twice.
Or we can make it a multi-stream, where we treat each listen as a new wrapping of the original. 
Then each listener on the new stream sees a result computed from the events it would have seen if subscribing to the original stream at the same time. Because that's precisely what happens.
There is no interaction between them because they are not all looking at the same broadcast stream.



I guess the underlying problem is that I'm not sure what the user intent is for allowing multiple listeners on a stream.

A broadcast stream is fire-and-forget, listeners come and go. They are usually considered started before anyone listens, it's just that they know not to do anything. 

A single-subscription stream is intended for giving all its data to one listener, in order, starting when listened to.

It's not clear whether doing, say, `.throttle` is intended to *create* a broadcast or single-subscription stream.
It's definitely not obvious that it should always be the same as the source stream, because what you want depends on the use, not the source.
It's easier for something like `.map` where mapping the original events once or once per listener only affects how many times the `convert` function is called, not what the events are (assuming the `convert` function is side-effect free).

Any *history* based operation applied to a non-single-subscription stream can be treated in two ways:
* Use the global history and give everybody the same result.
* Use the individual history for each listener, only based on events emitted while it is listening.

The former either means listening immediately (or sooner?!) or having the result depend on when the first listener happens. The former is potentially expensive and wasteful. The latter is just weird and arbitrary.

And it's not clear why that behavior should only apply to broadcast source streams. 
It's not clear that the source being a broadcast stream is important, after all, you only listen to it once anyway, so why not do the same thing for all streams?

I guess it's just a weird model to apply a history-based transformation to something that *might* be running before you start. You should always assume that a stream does nothing until you listen, and you should (preferably, if possible) not listen until you know events are needed.

The best model for that is to not create a broadcast stream.

Creating a single-subscription stream is always valid. You can only use it once, but if the source is not a single-subscription stream, you can always create another. 
It's safe and understandable.

(And creating a multi-stream is just removing that one extra step, and allowing you to treat each `listen` as a new listen on a separate stream).

Reply:
> It's definitely not obvious that it should always be the same as the source stream, because what you want depends on the use, not the source.

I think the choice to maintain the broadcast status of the source stream was fairly arbitrary. I wonder if it would make sense to overhaul all the APIs to return single subscription streams consistently, and force callers to use `asBroadcastStream` if they want to get back to a broadcast stream.

Reply:
I think there are APIs where it makes more sense to preserve broadcast-ness, like `map` or `where`, where (assuming pure functions), each listener will see the same thing, because the result depends entirely on the individual events emitted while you listen.

Synchronous per-event conversions, basically (to zero (`where`), one (`map`) or multiple (`expand`) new events, but all delivered immediately), where everything happens immediately when the original event is emitted, with no history or delay.

Even `asyncExpand` on `Stream` is not one of those, and I think it currently preserves broadcast-ness.

I'd consider preserving broadcast-ness for those, but it's also fair to use a single-sub stream and ask you to convert the stream twice which duplicates the computation for each listener. (That was actually a design goal for Dart 2.0, which slipped.) 

Or use a multi-stream, which acts like you did wrap it twice. 
Then I'd also use a multi-streams on all non-broadcast streams, in case the source is itself a multi-stream, so listening twice will work.

Always using a single-sub stream or a multi-stream is the simplest approach. Preserving broadcast-ness is really just an optimization.